File: src/com/googlecode/utterlyidle/ConsumesMimeMatcher.scala
package org.webfabric.rest

import java.lang.reflect.Method
import javax.ws.rs.{Consumes}
import javax.ws.rs.core.{HttpHeaders, MediaType}

class ConsumesMimeMatcher(method: Method) extends Matcher[Request] {
  lazy val mimeType: String = {
    List(method.getAnnotation(classOf[Consumes]), method.getDeclaringClass.getAnnotation(classOf[Consumes])).filter(_ != null) match {
      case x :: xs => x.value.first
      case Nil => MediaType.WILDCARD
    }
  }

  def isMatch(request: Request): Boolean = {
    if (mimeType.equals(MediaType.WILDCARD)) true else
    if (request.headers.contains(HttpHeaders.CONTENT_TYPE)) request.headers.getValue(HttpHeaders.CONTENT_TYPE).equals(mimeType) else true
  }
}

File: src/com/googlecode/utterlyidle/ConsumesMimeMatcher.java
package com.googlecode.utterlyidle;

import com.googlecode.totallylazy.Predicate;
import com.googlecode.totallylazy.Sequence;
import com.googlecode.totallylazy.predicates.LogicalPredicate;

public class ConsumesMimeMatcher implements Predicate<Request> {
    private final Sequence<String> mimeTypes;

    public ConsumesMimeMatcher(Sequence<String> mimeTypes) {
        this.mimeTypes = mimeTypes;
    }

    public boolean matches(final Request request) {
        if (mimeTypes.contains(MediaType.WILDCARD)) return true;
        return request.headers().valueOption(HttpHeaders.CONTENT_TYPE).exists(new LogicalPredicate<String>() {
            @Override
            public boolean matches(final String type) {
                return mimeTypes.exists(new Predicate<String>() {
                    public boolean matches(String mimeType) {
                        return type.startsWith(mimeType);
                    }
                });
            }
        });
    }

    public static Predicate<Binding> contentMatches(final Request request) {
        return new Predicate<Binding>() {
            public boolean matches(Binding binding) {
                return new ConsumesMimeMatcher(binding.consumes()).matches(request);
            }
        };
    }

}